Generalize cube extract

kecnry opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request generalizes cubeviz's spectral extraction so that it can be re-used by lcviz (photometric extraction of target pixel files) and eventually rampviz.  This is milestoned for the 4.0 release, which already contains breaking changes to spectral extraction.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

rosteen commented on jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:
```diff
@@ -198,6 +187,19 @@ def user_api(self):
+        else:
+            # no cube-like data loaded.  Once loaded, the parser will unset this
+            self.disabled_msg = (
+                f"{self.__class__.__name__} requires a single dataset to be loaded, "
+                "please load data to enable this plugin."
```
This would be a good time to update this to be more informative/correspond better with what is actually being checked. I'd change it to mention that the plugin is expecting a _three dimensional_ dataset to be loaded.

kecnry replied:
Good point, I'll make that change.

rosteen reviewed (approved):
One minor suggestion about some disabled message text, but other than that I think this looks good - less complicated that the diff suggested, mostly just moving things around and renaming.

bmorris3 commented on jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:
```diff
@@ -198,6 +187,19 @@ def user_api(self):
+    @observe('dataset_items')
+    def _update_disabled_msg(self, msg={}):
+        for data in self.app.data_collection:
+            if len(data.data.shape) == 3:
```
```suggestion
            if data.ndim == 3:
```

bmorris3 commented on jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py:
```diff
@@ -209,11 +211,15 @@ def __call__(self, add_data=True):
         self.aperture._set_mark_visiblities(self.active_step in ('', 'ap', 'ext'))
         self.background._set_mark_visiblities(self.active_step == 'bg')
-        self.marks['bg_spec'].visible = self.active_step == 'bg'
+        self.marks['bg_ext'].visible = self.active_step == 'bg'
```
Is `ext` for "extraction", rather than its usual "extension" as abbreviated elsewhere in jdaviz? This would require changes throughout the PR, but could we use a less degenerate abbreviation? Maybe `extr`?

bmorris3 replied:
(While I'm aware that I can be a pedant,) more descriptive names are always better.

bmorris3 reviewed (commented):
Some superficial comments for readability.

kecnry commented:
@bmorris3 / @rosteen - added a few more commits since your last reviews if you want to re-review (should be able to filter the diff to just the last ~6 commits)

@bmorris3 - addressed your comments.  I haven't touched renaming background or wavelength-dependent related methods and attributes that don't need to be used for lcviz (`wavelength_dependent`, `bg_wavelength_dependent`, `bg_spec_per_spaxel`, `bg_spec_add_results`, `extract_bg_spectrum`) - but will rampviz need access to these and maybe we should generalize them (either here or as a separate effort before they get released in 4.0)?

rosteen commented:
I seem to be getting some looping behavior now, could this be due to removing `skip_if_no_updates_since_last_active`? It did stop eventually but took a lot longer than I would expect. This was after selecting a spatial subset for a sum extraction.

https://github.com/user-attachments/assets/5b815992-0fce-436d-b63e-f9ec4c8d5a07

kecnry commented:
drats, I wasn't seeing that but that is most likely the cause.  I guess I'll have to split that method after all.

bmorris3 commented on jdaviz/core/template_mixin.py:
```diff
@@ -2442,7 +2443,9 @@ def _update_mark_coords(self, *args):
-                 spectral_display_unit, reference_spectral_value=None):
+                 slice_display_unit, spatial_axes=(0, 1), reference_spectral_value=None):
+        # slice_axis is the remaining axis (of (0, 1, 2)) not included in spatial_axes
+        slice_axis = 3 - sum(spatial_axes)
```
I was kind of mystified that this line works. I checked that it indeed works, but it feels like numerology? I've put in an equivalent but maybe less coincidental implementation below for your consideration.

```python
spatial_axes = [(0, 1), (1, 2), (0, 2)]
answers = [2, 0, 1]
for input, output in zip(spatial_axes, answers):

    assert 3 - sum(input) == output
    
    # suggested alternative:
    assert set([0, 1, 2]).difference(input).pop() == output
```

```suggestion
        slice_axis = set([0, 1, 2]).difference(input).pop()
```

kecnry replied:
I can see how that's a little more obvious to read (although I do think the sum logic makes sense and isn't just a coincidence - it just only works when you want to find the single missing number from a set), but probably less performant?

bmorris3 replied:
I've tested it, and of course it is less performant – but it's the difference between 40 nanoseconds for the original and 150 ns for my suggestion. I'm not sure anyone will notice. 😉

bmorris3 reviewed (approved):
Tested it out and it looks great. Thanks!

rosteen commented:
In addition to the preview not updating properly as reported offline, I'm a little worried that the extraction calculation is not correct either when a spatial subset is selected. In the screenshot below you can see that I selected a spatial subset that covers basically the whole spatial extent of the cube, but the extracted sum is significantly lower than the sum for the cube with no spatial subset.

<img width="805" alt="Screenshot 2024-07-15 at 10 51 41 AM" src="https://github.com/user-attachments/assets/a78fd027-1db2-476b-9dce-608c74e912f7">

kecnry commented:
@rosteen - is this different than on main?

rosteen commented:
Hmm, turns out it is present on main. Here you can see that even a subset covering the full spatial extent ends up with a smaller sum:

<img width="1323" alt="Screenshot 2024-07-15 at 11 12 49 AM" src="https://github.com/user-attachments/assets/a7ed5e6f-20cf-4f7b-ace3-15f363af8566">

rosteen commented:
See https://github.com/spacetelescope/jdaviz/issues/3086

kecnry commented:
@rosteen - can you take another look now and see if the live-previews now act as you'd expect?  Some of this buggy behavior was on main previously, but I do think this is a good chance to fix them, so would like to get it right if possible.

kecnry commented:
> I'm still a little surprised that moving/resizing a subset doesn't update the preview, but I think that's just because it's hitting the >1 second update time that triggers the warning and requires a manual click to update.

I can try to implement this (I think its the same on main - but agreed that it isn't correct).

> Related to that (but probably not related directly to this PR), I'm a little surprised that if you click "Disable Previews" when they're taking too long, there doesn't seem to be a way to ever turn them back on from the UI. Maybe I'm missing something, and this could be a follow up anyway.

They can be re-enabled from the "settings" section at the top of the plugin.  This is mentioned in the tooltip text, but maybe should be more prominent somewhere.  That feels like a separate effort to me that might need more input, so feel free to create a follow-up ticket if you have any ideas/suggestions.

kecnry commented:
>> I'm still a little surprised that moving/resizing a subset doesn't update the preview, but I think that's just because it's hitting the >1 second update time that triggers the warning and requires a manual click to update.

> I can try to implement this (I think its the same on main - but agreed that it isn't correct).

This is proving more difficult than I had hoped.  On main, it seems that the preview is just dropped when the subset is updated, so I don't think this is a reversion, but because the _visibility_ of the preview has been fixed here, this does expose this "bug".  Would you be ok with this as a follow-up ticket or do you think that because it exposes this bug more clearly (by fixing the visibility bug) that the bug should block this PR?

kecnry commented:
I think (again, hopefully for real this time) that I have the live-previews updating correctly for a change to the selected subset.  See the last ~5 commits for the diff since your previous reviews.  @rosteen @bmorris3 - please take a quick look/test and let me know if you have any comments or if your approvals still stand.

rosteen commented:
I think it looks ok, my approval can stand.

bmorris3 reviewed (approved):
Thanks @kecnry, looks good 👌🏻
